lua/lualine/components/branch.lua

```diff
@@ -1,33 +1,83 @@
-local async = require('lualine.async')
-
 local git_branch
 
-local get_git_branch = async:new({
-  cmd = 'git branch --show-current',
-  on_stdout = function(_, data)
-    if data then
-      git_branch = data:gsub('\n', '')
+local known_git_dirs = {}
+local sep = package.config:sub(1,1)
+
+local function exists(path)
+  local f = io.open(path)
+  if not f then return false, false end
+  local _, _, errno = f:read()
+  f:close()
+  return true, errno == 21 -- is a directory
+end
+
+local function find_git_dir()
+  -- path seperator is not in the end add it
+  local file_path = vim.fn.expand('%:p:h')..sep
+  -- don't show branch for terminal buffers
+  if file_path:match('^term://.*$') then return nil end
+
+  local dir = file_path
+  -- do we need to iterate more than that?
+  local n = 30
+  while dir and n > 0 do
+    n = n - 1
+    -- See if we already know where it is
+    if known_git_dirs[dir] then
+      if dir ~= file_path then
+        -- store it if file_path is a sub directory of dir
+        known_git_dirs[file_path] = known_git_dirs[dir] end
+      return known_git_dirs[dir]
     end
-  end,
-  on_stderr = function (_, data)
-    if data then
-      if data:find("fatal: not a git repository") then
-        git_branch = ''
+    local git_dir = dir..'.git'
+    local is_available, is_directory = exists(git_dir)
```

Comment:
this whole loop can be replaced with 
```lua
  local dir = file_path
  while dir do
    local git_dir = dir..'.git'
    local is_directory = vim.fn.isdirectory(git_dir)
    if is_directory==1 then
      return git_dir
    end
    dir = dir:match("(.*"..sep..").+$") -- "(.*/).+$"
  end
```

Reply:
please check if this does not produce any new issues or reproduce above mentioned ones. For the minimal testing I could do it works just fine

Reply:
actually there might be an even better way, check out `:h finddir` and `:h file-searching`

Reply:
try
```vim
:echo finddir('.git', '.;')
```
I think this is a drop in replacement for this whole function

Reply:
Just looking for dir wonn't work as when 
`git init --separate-dir` is used or it's a submodule . In those cases .git is a file. Plus I'm sure this lua implenentation will way more efficient then vimL implementation.

Reply:
And I am sure that `finddir` is not written in vimscript but in C :)

Reply:
>And I am sure that finddir is not written in vimscript but in C :)

Alright then it's a good point . But we still will have to handle those submodules . It still doesn't change the problem we are having with full path . as we will still have to handle it .

Reply:
full path? I think only submodules should be handled. I honestly have no idea how to do it and you seemed to have figured it out already so could you handle it? Maybe some smart if will do the trick.
Right now I am using 
```lua
  local git_dir = vim.fn.finddir('.git', '.;')
```
you can find the `.git` file easily with `findfile` but I have no idea on how to parse it

Reply:
I'm not sure but I think finddir and findfile runs from curent dir insted of filedir . can you tell me what '.;' means ? 
It reintruduces the issue where branch doesn't change on non git dir.

Reply:
Oh snap it does...

well `.;` can be translated to 
`.` vim starting directory
`;` until you find a desired file or reach the top of tree

I pulled this up in 3 minutes, there should be an option to replace `.` with current dir. If not, we can just use the expanded filepath because it also accepts relative & full paths

Reply:
yeah, rather easy fix, I think this works
```lua
local function watch_head()
  file_changed:stop()
  local dir = vim.fn.expand('%:p:h')..sep
  local git_dir = vim.fn.finddir('.git', dir .. ';')
  if #git_dir > 0 then
    local head_file = git_dir..sep..'HEAD'
    get_git_head(head_file)
    file_changed:start(head_file, {}, vim.schedule_wrap(function()
      -- reset file-watch
      watch_head()
    end))
  else
    -- set to nil when git dir was not found
    git_branch = nil
  end
end
```